chromium_src/third_party/blink/renderer/modules/storage/dom_window_storage.cc

```diff
@@ -8,47 +8,160 @@
-// it was caused by CanAccessSessionStorage for the document security origin
-// (sandbox, data urls, etc...)
-void MaybeClearAccessDeniedException(StorageArea* storage,
-                                     const LocalDOMWindow& window,
-                                     ExceptionState* exception_state) {
-  if (!storage && exception_state->HadException()) {
-    Document* document =
-        window.GetFrame()->GetDocument();
-
-    if (!document->GetSecurityOrigin()->CanAccessSessionStorage())
-      return;
-
-    // clear the access denied exception for better webcompat
-    exception_state->ClearException();
-  }
+class EphemeralStorageNamespaces
+    : public GarbageCollected<EphemeralStorageNamespaces>,
+      public Supplement<Page> {
+  USING_GARBAGE_COLLECTED_MIXIN(StorageNamespace);
+
+ public:
+  EphemeralStorageNamespaces(StorageController* controller,
+                             const String& session_storage_id,
+                             const String& local_storage_id);
+  virtual ~EphemeralStorageNamespaces() = default;
+
+  static const char kSupplementName[];
+  static EphemeralStorageNamespaces* From(Page* page);
+
+  StorageNamespace* session_storage() { return session_storage_.Get(); }
+  StorageNamespace* local_storage() { return local_storage_.Get(); }
+  void Trace(Visitor* visitor) const override;
+
+ private:
+  Member<StorageNamespace> session_storage_;
+  Member<StorageNamespace> local_storage_;
+};
+
+const char EphemeralStorageNamespaces::kSupplementName[] =
+    "EphemeralSessionStorageNamespaces";
+
+EphemeralStorageNamespaces::EphemeralStorageNamespaces(
+    StorageController* controller,
+    const String& session_storage_id,
+    const String& local_storage_id)
+    : session_storage_(
+          MakeGarbageCollected<StorageNamespace>(controller,
+                                                 session_storage_id)),
+      local_storage_(MakeGarbageCollected<StorageNamespace>(controller,
+                                                            local_storage_id)) {
+}
+
+void EphemeralStorageNamespaces::Trace(Visitor* visitor) const {
+  visitor->Trace(session_storage_);
+  visitor->Trace(local_storage_);
+  Supplement<Page>::Trace(visitor);
 }
 
-}  // namespace
+// static
+EphemeralStorageNamespaces* EphemeralStorageNamespaces::From(Page* page) {
+  if (!page)
+    return nullptr;
+
+  EphemeralStorageNamespaces* supplement =
+      Supplement<Page>::From<EphemeralStorageNamespaces>(page);
+  if (supplement)
+    return supplement;
+
+  // The ephemeral session storage namespace is constructed using the
+  // normal session storage id. We get it here and transform it.
+  StorageNamespace* session_storage_namespace = StorageNamespace::From(page);
+  if (!session_storage_namespace)
+    return nullptr;
 
-BraveDOMWindowStorage::BraveDOMWindowStorage() {}
-BraveDOMWindowStorage::~BraveDOMWindowStorage() {}
+  String session_storage_id = session_storage_namespace->namespace_id() +
```

Comment:
I don't understand the point of creating these new namespaces, why are we just using the existing one?

Reply:
In this case we are creating namespaces for `localStorage` and `sessionStorage` if the `EphemeralStorageNamespaces` hasn't been been added as a supplement to the `Page` yet. We don't have an existing one apart from the namespace used for non-ephemeral-`sessionStorage`, which cannot be used for the ephemeral version. A `Page` may contain both types.

Reply:
`sessionStorage` is ephemeral and the new namespaces you are always (correctly) creating these with `kSessionStorage` just like the original `sessionStorage`

Reply:
basically `sessionStorage` should stay as normal `sessionStorage`, but with the additional caveat that we will clear it on main frame tld change. We only need a new namespace for `localStorage` so it can act like `sessionStorage` in this context

Reply:
If we clear `sessionStorage` we will also be clearing any `sessionStorage` for the first-party domain. Essentially ephemeral storage can always co-exist in a tab with non-ephemeral storage. We also can't track third-party domains and only clear those, because if you visit

 1. a.com
 2. b.com (a.com in an iframe)
 3. a.com

we want a.com's `sessionStorage` to be preserved between 1 and 3, but to be isolated from a.com's session storage in 2.

Reply:
it should only be cleared for this namespace which is unique per tab

Reply:
and that scenario is incorrect. When this feature is enabled, if the eTLD changes the sessionStorage should be cleared

Reply:
I guess there was a bit of a misunderstanding between myself and @pes10k, so I understand now why you're separating it from the normal sessionStorage